File: examples/rust/src/tuna_position/close_position_with_liquidity_orca.rs
use crate::{types::Amounts, utils::rpc::create_and_send_transaction};
use anyhow::Result;
use defituna_client::{
  accounts::{fetch_market, fetch_tuna_position},
  close_position_with_liquidity_orca_instructions, get_market_address, get_tuna_position_address,
  ClosePositionWithLiquidityOrcaArgs,
};
use solana_client::rpc_client::RpcClient;
use solana_sdk::{pubkey::Pubkey, signer::Signer};

/// Removes liquidity from a position in an Orca liquidity pool and closes it, managing funds via Tuna's smart contract.
///
/// This function uses the SOL/USDC Whirlpool with preset withdrawal amounts and swap options. These presets can be adjusted or passed through the function’s input.
///
/// **Note:** This function combines opening and removing liquidity, though these actions can be performed separately. Liquidity can be removed multiple times based on the available liquidity in the Orca position.
///
/// # Parameters
/// - `rpc`: The `RpcClient` for interacting with the Solana blockchain.
/// - `authority`: The authority `Signer` who owns the position.
/// - `tuna_position_mint`: The `Pubkey` of the Tuna Position Mint identifying the position to manage.
///
/// # Returns
/// - `Result<()>`: Returns `Ok(())` if the transaction is successful, or an error if it fails.
pub fn close_position_with_liquidity_orca(
  rpc: RpcClient,
  authority: Box<dyn Signer>,
  tuna_position_mint: Pubkey,
) -> Result<()> {
  // The Tuna Position Account containing deserialized data, fetched using Tuna's Client.
  let tuna_position = fetch_tuna_position(&rpc, &get_tuna_position_address(&tuna_position_mint).0)?;

  // The Market Account containing deserialized data, fetched using Tuna's Client.
  let market = fetch_market(&rpc, &get_market_address(&tuna_position.data.pool).0)?;

  // Minimum removed amounts for Tokens A and B to be respected by the RemoveLiquidity instruction, acting as slippage limits.
  let min_removed_amount = Amounts { a: 0, b: 0 };
  // The total amount of slippage allowed on the Whirlpool's price during potential inner swaps due to deposit ratio rebalancing.
  let max_swap_slippage = 0;
  // The option for whether to swap and which token to swap to during RemoveLiquidity.
  // - 0 - No swap
  // - 1 - Swaps to Token A
  // - 2 - Swaps to Token B
  let swap_to_token = 1;

  // Creation of instructions for removing liquidity and closing positions;
  let args = ClosePositionWithLiquidityOrcaArgs {
    swap_to_token,
    min_removed_amount_a: min_removed_amount.a,
    min_removed_amount_b: min_removed_amount.b,
    max_swap_slippage,
  };

  let mut instructions =
    close_position_with_liquidity_orca_instructions(&rpc, &authority.pubkey(), &tuna_position_mint, args)?;

  // Signing and sending the transaction with all the instructions to the Solana network.
  create_and_send_transaction(
    &rpc,
    &authority,
    &mut instructions,
    None,
    Some(market.data.address_lookup_table),
    None,
  )
}

File: examples/ts/src/tunaPosition/closePositionWithLiquidityOrca.ts
import {
  ClosePositionWithLiquidityOrcaInstructionArgs,
  closePositionWithLiquidityOrcaInstructions,
  fetchMarket,
  getMarketAddress,
} from "@crypticdot/defituna-client";
import { Address, address, IInstruction } from "@solana/kit";
import { SOL_USDC_WHIRLPOOL } from "src/constants";
import { loadKeypair } from "src/utils/common";
import { createAndSendTransaction, rpc } from "src/utils/rpc";

/**
 * Removes liquidity from a position in an Orca *Liquidity Pool* and closes it, managing funds via Tuna's smart contract.
 * Uses the SOL/USDC *Whirlpool* with preset withdrawal amounts and swap options for this example; these can be adjusted or passed through the function’s input.
 * Note: Combines opening and removing liquidity, though these actions can be performed separately, and liquidity can be removed multiple times
 * based on the available liquidity in the *Orca Position*.
 * @param {Address} tunaPositionMint - The {@link Address address} of the *Tuna Position Mint* identifying the position to manage.
 * @returns {Promise<void>} A promise that resolves when the transaction is confirmed.
 * @throws {Error} If the transaction fails to send or confirm.
 */
export async function closePositionWithLiquidityOrca(tunaPositionMint: Address): Promise<void> {
  /**
   * Defining input variables.
   */
  const whirlpoolAddress = SOL_USDC_WHIRLPOOL;
  /**
   * Minimum removed amounts for Tokens A and B to be respected by the *RemoveLiquidity* instruction, acting as slippage limits.
   */
  const minRemovedAmount = { a: 0n, b: 0n };
  /**
   * The total amount of slippage allowed on the {@link Whirlpool Whirlpool}'s `price` during potential inner swaps due to deposit ratio rebalancing.
   */
  const maxSwapSlippage = 0;
  /**
   * The option for whether to swap and which *token* to swap to during *RemoveLiquidity*.
   * - 0 - No swap
   * - 1 - Swaps to *Token A*
   * - 2 - Swaps to *Token B*
   */
  const swapToToken = 1;

  //
  // Creation of instructions for removing liquidity and closing the position.
  //

  const authority = await loadKeypair();
  const market = await fetchMarket(rpc, (await getMarketAddress(whirlpoolAddress))[0]);

  const instructions: IInstruction[] = [];

  const args: ClosePositionWithLiquidityOrcaInstructionArgs = {
    swapToToken,
    minRemovedAmountA: minRemovedAmount.a,
    minRemovedAmountB: minRemovedAmount.b,
    maxSwapSlippage,
  };

  instructions.push(...(await closePositionWithLiquidityOrcaInstructions(rpc, authority, tunaPositionMint, args)));

  /**
   * Signing and sending the transaction with all the instructions to the Solana network.
   */
  await createAndSendTransaction(authority, instructions, market.data.addressLookupTable);
}

const tunaPositionMint = process.argv[2];
if (!tunaPositionMint) {
  console.error("Please provide the address for the tunaPositionMint as an argument.");
  process.exit(1);
}

closePositionWithLiquidityOrca(address(tunaPositionMint)).catch(console.error);
